pytorch_lightning/trainer/connectors/data_connector.py

```diff
@@ -70,21 +78,37 @@ def get_profiled_train_dataloader(self, train_dataloader) -> Iterable:
     def prepare_data(self) -> None:
         # on multi-gpu jobs we only want to manipulate (download, etc) on node_rank=0, local_rank=0
         # or in the case where each node needs to do its own manipulation in which case just local_rank=0
-        if self.can_prepare_data():
-            if self.trainer.datamodule is not None:
-                self.trainer.datamodule.prepare_data()
+        should_call_once_per_node = self.trainer.local_rank == 0
+        should_call_once_in_total = self.trainer.local_rank == 0 and self.trainer.node_rank == 0
+
+        # handle datamodule prepare data:
+        # check for prepare_data_per_node & datamodule lifecycle properties before calling datamodule.prepare_data
+        if self.trainer.datamodule is not None:
+            if is_overridden("prepare_data", self.trainer.datamodule) and (
+                not self.trainer.datamodule.has_prepared_data
+            ):
+                dm_prepare_data_per_node = (
+                    self.trainer.prepare_data_per_node
+                    if self.trainer.datamodule is None
+                    else self.trainer.datamodule.prepare_data_per_node
+                )
+                if (dm_prepare_data_per_node and should_call_once_per_node) or (
+                    not dm_prepare_data_per_node and should_call_once_in_total
+                ):
+                    self.trainer.datamodule.prepare_data()
+        # handle lightning module prepare data:
+        # check for prepare_data_per_node before calling lightning_module.prepare_data
+        lm_prepare_data_per_node = (
+            self.trainer.prepare_data_per_node
+            if self.trainer.lightning_module is None
+            else self.trainer.lightning_module.prepare_data_per_node
+        )
```

Comment:
this should prioritize the field on the lightning module instead of the trainer

`self.trainer.lightning_module.prepare_data_per_node` should take priority over `self.trainer.prepare_data_per_node` because this is the way we can phase out `trainer.prepare_data_per_node` over time

Reply:
I can rephrase this line a little bit, but overall, as long as `self.trainer.lightning_module` exists, we use the `prepare_data_per_node ` from `lightning_module `.